cupyx/scipy/sparse/linalg/_iterative.py

```diff
@@ -93,6 +97,122 @@ def cg(A, b, x0=None, tol=1e-5, maxiter=None, M=None, callback=None,
+            relative (preconditioned) residual norm is used as an arugment.
+        atol (float): Tolerance for convergence.
+
+    Returns:
+        tuple:
+            It returns ``x`` (cupy.ndarray) and ``info`` (int) where ``x`` is
+            the converged solution and ``info`` provides convergence
+            information.
+
+    Reference:
+        M. Wang, H. Klie, M. Parashar and H. Sudan, "Solving Sparse Linear
+        Systems on NVIDIA Tesla GPUs", ICCS 2009 (2009).
+
+    .. seealso:: :func:`scipy.sparse.linalg.gmres`
+    """
+    if A.ndim != 2 or A.shape[0] != A.shape[1]:
+        raise ValueError('expected square matrix (shape: {})'.format(A.shape))
+    if A.dtype.char not in 'fdFD':
+        raise TypeError('unsupprted dtype (actual: {})'.format(A.dtype))
+    n = A.shape[0]
+    if not (b.shape == (n,) or b.shape == (n, 1)):
+        raise ValueError('b has incompatible dimensins')
+    b = b.astype(A.dtype).ravel()
+    if n == 0:
+        return cupy.empty_like(b), 0
+    b_norm = cupy.linalg.norm(b)
+    if b_norm == 0:
+        return b, 0
+    if atol is None:
+        atol = tol * float(b_norm)
+    else:
+        atol = max(float(atol), tol * float(b_norm))
+    if x0 is None:
+        x = cupy.zeros((n,), dtype=A.dtype)
+    else:
+        if not (x0.shape == (n,) or x0.shape == (n, 1)):
+            raise ValueError('x0 has incompatible dimensins')
+        x = x0.astype(A.dtype).ravel()
+    if maxiter is None:
+        maxiter = n * 10
+    if restart is None:
+        restart = 20
+    restart = min(restart, n)
+    if callback_type is None:
+        callback_type = 'pr_norm'
+    if callback_type not in ('x', 'pr_norm'):
+        raise ValueError('Unknow callback_type: {}'.format(callback_type))
+    if callback is None:
+        callback_type = None
+
+    V = cupy.empty((n, restart), dtype=A.dtype, order='F')
+    H = cupy.zeros((restart+1, restart), dtype=A.dtype, order='F')
+    e = numpy.zeros((restart+1,), dtype=A.dtype)
+
+    matvec, psolve = _make_funcs(A, M)
+    compute_hu = _make_compute_hu(V)
+
+    iters = 0
+    while True:
+        mx = psolve(x)
+        r = b - matvec(mx)
+        r_norm = cublas.nrm2(r)
+        if callback_type == 'x':
+            callback(mx)
+        elif callback_type == 'pr_norm' and iters > 0:
+            callback(r_norm / b_norm)
+        if r_norm <= atol or iters >= maxiter:
+            break
+        v = r / r_norm
+        V[:, 0] = v
+        e[0] = r_norm
+
+        # Arnoldi iteration
+        for j in range(restart):
+            z = psolve(v)
+            u = matvec(z)
+            H[:j+1, j], u = compute_hu(u, j)
+            cublas.nrm2(u, out=H[j+1, j])
+            if j+1 < restart:
+                v = u / H[j+1, j]
+                V[:, j+1] = v
+
+        # Note: The least-square solution to equation Hy = e is computed on CPU
+        # because it is faster if tha matrix size is small.
+        ret = scipy.linalg.lstsq(cupy.asnumpy(H), e)
```

Comment:
Can we use `numpy.linalg.lstsq` instead as CuPy does not require scipy.

Reply:
Ah, scipy is not available be default, isn't it. I will change it to use `numpy.linalg.lstsq`.